rimage/manifest.c

```diff
@@ -652,6 +665,84 @@ static int man_hash_modules(struct image *image, struct sof_man_fw_desc *desc)
 	return 0;
 }
 
+/* used by others */
+static int man_write_fw_v1_5(struct image *image)
+{
+	struct sof_man_fw_desc *desc;
+	struct fw_image_manifest_v1_5 *m;
+	int ret;
+
+	/* init image */
+	ret = man_init_image_v1_5(image);
+	if (ret < 0)
+		goto err;
+
+	/* open ROM image */
+	ret = man_open_rom_file(image);
+	if (ret < 0)
+		goto err;
+
+	/* open unsigned firmware */
+	ret = man_open_unsigned_file(image);
+	if (ret < 0)
+		goto err;
+
+	/* create the manifest */
+	ret = man_open_manifest_file(image);
+	if (ret < 0)
+		goto err;
+
+	/* create the module */
+	m = image->fw_image;
+	desc = image->fw_image + MAN_DESC_OFFSET;
+
+	/* create each module */
+	m->desc.header.num_module_entries = image->num_modules;
+	man_create_modules(image, desc);
+
+	fprintf(stdout, "Firmware completing manifest v1.5\n");
+
+	/* create structures from end of file to start of file */
+	ri_adsp_meta_data_create(image, MAN_META_EXT_OFFSET_V1_8,
+				 MAN_FW_DESC_OFFSET_V1_8);
+	ri_plat_ext_data_create(image);
+	ri_css_hdr_create_v1_8(image);
```

Comment:
v1.8 header and offsets in v1.5 func ?

Reply:
SInce the ri_css_hdr_create_v1_8 is only to set the date of the header, I will change this API to ri_css_hdr_create 1.8 and 1.5 all using this API.

Reply:
@zhuyingjiang the important thing is that the structure (and some of the contents) of the file should look exactly like the existing files. i.e.
```
hexdump -C /lib/firmware/intel/dsp_fw_kbl.bin | less

00000000  24 41 45 31 48 05 00 00  01 00 00 00 0c 00 00 00  |$AE1H...........|
00000010  28 00 00 00 b9 0c eb 61  d8 34 59 4f a2 1d 04 c5  |(......a.4YO....|
...
00000540  00 00 00 00 00 00 00 00  06 00 00 00 a1 00 00 00  |................|
00000550  00 00 01 00 00 00 00 00  86 80 00 00 08 01 18 20  |............... |
00000560  00 08 00 00 40 00 00 00  40 00 00 00 01 00 00 00  |....@...@.......|
00000570  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
000005c0  00 00 00 00 00 00 00 00  0b 93 f6 fd 4f 5a de 80  |............OZ..|
...
000007c0  5d ab 07 25 db 32 0e 70  a9 71 b5 91 24 41 4d 31  |]..%.2.p.q..$AM1|
000007d0  34 00 00 00 41 44 53 50  46 57 00 00 3b 00 00 00  |4...ADSPFW..;...|
000007e0  00 00 00 00 1f 00 00 00  09 00 15 00 00 00 c2 0c  |................|
000007f0  0c 00 00 00 00 00 05 be  00 a0 04 00 00 00 00 00  |................|
```
The first section to 0x548 is extended manifest and is not needed by SOF. The CSS header starts at 0x548, this should be offset 0x0 in SOF. ADSP header starts at 0x7cc but should start at 0x284 in SOF. FW code starts at 0x2548 and should start at 0x2000 in SOF for manifest v1.5

Reply:
OK, will check it.

Reply:
Yes, they are the same:
Cavs1.5_Manifest:
        CssHeader:
            module_type: 6 -> LT_MODULE_TYPE
            header_len: 161
            header_version: 0x10000
            CssModuleId:
                reserved: 0
                debug_manifest: 0
            module_vendor: 0x8086
            date: 0
            size: 0x800
            key_size: 64
            modulus_size: 64
            exponent_size: 1
            reserved:
                 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000
                 0x00000000 0x00000000 0x00029b80 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000
                 0x00000000 0x20181101 0x00000000 0x00000000 0x00000000 0x00000000
        ManifestCryptoBlock: